### ISSUE TITLE
Non-deterministic output on repeated inference: same wrapper, same image, different signal (image-dependent)

### ISSUE THREAD

**Issue by paramson** [Ahus-AIM/Open-ECG-Digitizer#48]
### Summary

Running the same image through the same `InferenceWrapper` instance twice can
change the digitised signal by up to **1.4 mV across 7 of 12 leads**. In our
downstream use the **findings-level output was unaffected** — we feed the
digitiser into a deterministic rule engine, and across every comparison we made
its findings, priority flags and "insufficient" reasons were identical — so what
we are reporting is sub-finding numeric instability, not a wrong clinical
result. We are reporting it because our downstream criteria include voltage
thresholds (Sokolow-Lyon, Cornell), where a 10–17% amplitude swing could cross
one.

The effect is **image-dependent**: some images are perfectly bit-reproducible,
others are not, and the same images are the unstable ones across separate
processes too. It shows up in per-lead amplitudes and, on one image, in the
layout matching cost.

We think we have found the amplification point in `signal_extractor.py` and
suggest a small order-independent fix below. We have **not** identified the
upstream source of the perturbation.

### Environment

- Open-ECG-Digitizer commit `97a1508` (branch `main`)
- Weights: `unet_weights_07072025.pt`, `lead_name_unet_weights_07072025.pt`
- Python 3.12.10, torch 2.14.0+cpu, numpy 2.5.3, scipy 1.18.1
- Windows 11 (10.0.26200), CPU only
- Config derived from the shipped example. Relevant values:
  `device: cpu`, `resample_size: 3000`, `rotate_on_resample: true`,
  `apply_dewarping: false`, `PERSPECTIVE_DETECTOR.num_thetas: 250`,
  `CROPPER.granularity: 80`, `CROPPER.percentiles: [0.02, 0.98]`,
  `PIXEL_SIZE_FINDER: {min_number_of_grid_lines: 30, max_number_of_grid_lines: 70,
  lower_grid_line_factor: 0.3}`. Layout set restricted to `3x4+1R` and
  `standard_3x4`.

Note `apply_dewarping: false`, so the dewarper never runs in any of this.

### Reproduction

```python
"""Same wrapper instance, same image, twice. Run from the repo root."""
import sys
import numpy as np

from src.config.default import get_cfg
from src.utils import find_config_path, import_class_from_path
from src.digitize import decode_and_prepare_image, canonical_from_got_values

cfg_name, img = sys.argv[1], sys.argv[2]
cfg = get_cfg(find_config_path(cfg_name))
wrapper = import_class_from_path(cfg.MODEL.class_path)(**cfg.MODEL.KWARGS)

outs, costs = [], []
for i in (1, 2):
    got = wrapper(decode_and_prepare_image(img), layout_should_include_substring=None)
    arr = canonical_from_got_values(got).squeeze().cpu().numpy().astype(np.float64)
    cost = got.get("signal", {}).get("layout_matching_cost", float("nan"))
    outs.append(arr); costs.append(cost)
    print(f"call {i}: layout={got.get('layout_name')} cost={cost!r}")

a, b = outs
same = np.array_equal(np.nan_to_num(a, nan=-9e9), np.nan_to_num(b, nan=-9e9))
if same:
    print("RESULT: BIT-IDENTICAL within process")
else:
    d = np.abs(np.nan_to_num(a) - np.nan_to_num(b))
    rows = np.where(d.max(axis=1) > 0)[0]
    print(f"RESULT: DIFFERS  max|delta| = {d.max():.1f} uV  "
          f"leads differing: {len(rows)}/{a.shape[0]}  "
          f"cost {costs[0]:.6f} -> {costs[1]:.6f}")
```

Input images were 12-lead `3x4+1R` printout images, upscaled 2x with PIL LANCZOS
before digitising (originals were ~1200x600, below the resolution the pipeline
handles well).

The two unstable images are from the LITFL ECG Library (publicly available
educational ECGs — linked, not attached):

- **LBBB + AF** (the "DIFFERS, 1445 µV" record): Example 2 at
  <https://litfl.com/left-bundle-branch-block-lbbb-ecg-library/>
  (captioned "LBBB with AF")
- **Atrial fibrillation** (the "DIFFERS, 199 µV" record in Observation 2):
  Example 4 at <https://litfl.com/atrial-fibrillation-ecg-library/>
  ("AF with rapid ventricular response")

The bit-reproducible images were further examples from the same library
(normal sinus rhythm, first-degree AV block). One de-identified clinical
printout photo was also used; it is not included and cannot be shared. Any
12-lead 3x4+1R printout image should serve as a stable control.

### Observation 1 — within-process, same wrapper instance

| Image | Result | max &#124;Δ&#124; | Leads differing | Layout cost |
|---|---|---|---|---|
| LITFL LBBB + AF | **DIFFERS** | 1445.2 µV | 7 / 12 | unchanged (0.127516) |
| LITFL normal sinus rhythm | bit-identical | — | 0 / 12 | unchanged |
| LITFL 1st-degree AV block | bit-identical | — | 0 / 12 | unchanged |
| De-identified clinical photo | **DIFFERS** | 1970.6 µV | 12 / 12 | **0.281259 → 0.362837** |

Two points worth separating:

1. The effect is image-dependent. Two of the four images are perfectly
   reproducible on the second call, so this is not a blanket "second call is
   contaminated" behaviour.
2. On the clinical photo the **layout matching cost degraded** on call 2, while
   on the LITFL LBBB/AF image the cost was unchanged despite a 1.4 mV signal
   difference. We cannot tell from this whether we are seeing state carried
   between calls, or a computation that is non-deterministic and only visible on
   images sitting near a decision boundary. We did not try to distinguish these.

### Observation 2 — across separate processes

Same image, separate process invocations, identical config:

- One clinical printout photo, 5 runs: 10 of 12 leads bit-identical, and **all**
  derived interval measurements (rate, RR CV, PR, QRS, QT, QTc, axis) identical
  every run. Only aVL (16.8% peak-to-peak spread) and aVF (6.5%) moved. Layout
  matching cost identical to the last digit in all 5 runs.
- Same image with `OMP_NUM_THREADS=1` and `MKL_NUM_THREADS=1`, 2 runs: aVL still
  moved (1.32% spread, max &#124;Δ&#124; 13.8 µV). **Restricting threading does not
  eliminate it.**
- 6 LITFL images, batch mode (one wrapper across all six) vs one image per
  process: 4 of 6 records bit-identical. Only the AF image (max 199.1 µV) and
  the LBBB/AF image (max 631.8 µV) differed — and they differed regardless of
  their position in the batch. The LBBB/AF image was the first image processed
  in *both* modes and still differed.

The same images are unstable in both observations, which is what makes us think
this is one phenomenon rather than two.

### What we ruled out

By inspection and by test, not by assumption:

- **No RNG in the inference path.** Every `random` / `np.random` use is in
  training or augmentation code (`src/transform/vision.py`,
  `src/dataset/synthetic_lead_text.py`, `src/train.py`) and is unreachable from
  `src.digitize`. The one seed call, `src/digitize.py:27`, seeds only the
  shuffle of the input file list. There is no seed knob to set.
- **Models are in eval mode.** `@torch.no_grad()` at
  `src/model/inference_wrapper.py:80`, `.eval()` at `:315` and `:340` and at
  `src/model/lead_identifier.py:36,471`. No dropout/batch-norm stochasticity.
- **Threading** — tested, see above.
- **Uninitialised memory** — the `torch.empty` calls at
  `src/model/signal_extractor.py:54,58` and `src/model/lead_identifier.py:410`
  are degenerate zero-row returns, not read paths.
- **The dewarper** — `apply_dewarping: false`, so the branch at
  `src/model/inference_wrapper.py:117` never executes.
- **Per-call instance state** in `PerspectiveDetector`, `Cropper`,
  `PixelSizeFinder` and `LeadIdentifier` — these write `self` only in
  `__init__`. (`Dewarper` does keep substantial per-call state on `self`, but it
  is never called here.)

### Suspected amplification point

Regardless of what perturbs the input, we think this is where a tiny difference
becomes a visible one — `src/model/signal_extractor.py`:

```python
    def merge_components(self, lines, components):
        ...
        for group in components:
            group_lines = lines[torch.tensor(group)]          # line 312
            ...
            for col in range(group_lines.shape[1]):
                valid_values = group_lines[:, col][~torch.isnan(group_lines[:, col])]
                if len(valid_values) > 0:
                    merged_line[col] = valid_values[0]        # line 320
```

Where trace fragments overlap in a column, the merged value is
`valid_values[0]` — whichever fragment happens to come **first in the group's
ordering**, rather than a value-determined choice. That ordering is inherited
from `linear_sum_assignment` at line 274, through `build_match_graph` (line 277,
insertion order) and `get_connected_components` (line 286, `for node in graph`).

So any perturbation that changes the assignment — or the order it is emitted in
— changes which fragment's sample is kept, and the merged amplitude jumps
discretely. That matches what we see: the unstable images flip between a small
number of fixed values rather than drifting continuously, and the stable images
(cleanly separated traces, no overlapping fragments to choose between) never
move at all.

### Suggested fix

Make the resolution order-independent, so output is stable regardless of the
upstream ordering:

```python
                if len(valid_values) > 0:
                    merged_line[col] = valid_values.median()   # or .mean()
```

and/or make the group ordering itself deterministic:

```python
            group_lines = lines[torch.tensor(sorted(group))]
```

`median()` has the added benefit of being robust when one overlapping fragment
is an outlier, which is the case the current code is silently gambling on. A
maintainer will know better than us whether `valid_values[0]` encodes an
intentional "prefer the earliest-matched line" semantic — if it does, sorting
the group is the smaller change.

We have deliberately **not** submitted a patch, since we do not know the
intended semantics here and we cannot regression-test against your evaluation
set.

### Willing to help

We can run any suggested patch against our corpus (LITFL educational images plus
de-identified clinical printouts) and report per-lead spread across repeated
runs, if that is useful. Happy to provide the full per-lead numbers behind any
of the tables above.

---

Reported by an ED clinician building an educational ECG interpretation platform on top of this digitiser.
